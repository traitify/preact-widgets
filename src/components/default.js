import { h, Component } from "preact";

import Results from "./results";
import SlideDeck from "./slide-deck";

export default class Default extends Component {
  shouldLoadSlideDeck () {
    return (this.props.assessment.slides || []).length != 0;
  }
  shouldLoadResults () {
    return (this.props.assessment.personality_types || []).length != 0;
  }
  render() {
    let widget = <div />;
    if (this.shouldLoadSlideDeck()){
      widget = <SlideDeck {...this.props} />;
    }
    if (this.shouldLoadResults()){
      widget = <Results {...this.props} />;
    }
    return widget;
  }
}
