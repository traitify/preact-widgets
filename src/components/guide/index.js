import PropTypes from "prop-types";
import {Component} from "react";
import DangerousHTML from "lib/helpers/dangerous-html";
import TraitifyPropTypes from "lib/helpers/prop-types";
import withTraitify from "lib/with-traitify";
import CompetencySelect from "./competency-select";
import CompetencyTab from "./competency-tab";
import style from "./style.scss";

class Guide extends Component {
  static defaultProps = {assessment: null, guide: null}
  static propTypes = {
    assessment: PropTypes.shape({
      assessment_type: PropTypes.string,
      personality_types: PropTypes.array
    }),
    followGuide: PropTypes.func.isRequired,
    guide: PropTypes.shape({
      assessment_id: PropTypes.string.isRequired,
      competencies: PropTypes.array.isRequired,
      locale_key: PropTypes.string.isRequired
    }),
    isReady: PropTypes.func.isRequired,
    translate: PropTypes.func.isRequired,
    ui: TraitifyPropTypes.ui.isRequired
  }
  constructor(props) {
    super(props);
    this.state = {
      badges: [],
      competencies: [],
      displayedCompetency: null,
      showExpandedIntro: false
    };
  }
  componentDidMount() {
    this.props.ui.trigger("Guide.initialized", this);
    this.props.followGuide();
  }
  componentDidUpdate(prevProps) {
    this.props.ui.trigger("Guide.updated", this);

    const guide = this.props.guide || {};
    const prevGuide = prevProps.guide || {};
    const changes = {
      assessment: guide.assessment_id !== prevGuide.assessment_id,
      locale: guide.locale_key !== prevGuide.locale_key
    };

    if(changes.assessment || changes.locale) { this.setGuide(); }
  }
  setGuide() {
    if(!this.props.guide) {
      return this.setState({badges: [], competencies: [], displayedCompetency: null});
    }

    const {competencies} = this.props.guide;
    const badges = this.badges(competencies);

    this.setState({badges, competencies, displayedCompetency: {...competencies[0]}});
  }
  displayCompetency = (competencyID) => {
    this.setState(({competencies}) => {
      const displayedCompetency = competencies.find((competency) => competency.id === competencyID);

      return displayedCompetency && {displayedCompetency};
    });
  }
  tabBadge = (id) => (
    this.state.badges.find((badge) => badge.competencyID === id).image
  )
  toggleExpandedIntro = () => {
    this.setState((prevState) => ({showExpandedIntro: !prevState.showExpandedIntro}));
  }
  badges(competencies) {
    const {assessment} = this.props;

    return competencies.map((competency) => {
      const personalityID = competency.questionSequences[0].personality_type_id;
      const types = assessment.personality_types;
      const personality = types.find((type) => type.personality_type.id === personalityID);

      return {competencyID: competency.id, image: personality.personality_type.badge.image_medium};
    });
  }
  introduction() {
    const {introduction} = this.state.displayedCompetency;
    const intro = introduction.split("\n", 1)[0];
    const readMore = introduction.replace(intro, "").trim();

    return {intro, readMore};
  }
  stringToListItems(entity) {
    let entities = entity.split("\n");
    /* eslint-disable-next-line react/no-array-index-key */
    entities = entities.map((e, i) => <li key={i}>{e}</li>);

    return (<ul>{entities}</ul>);
  }
  render() {
    if(!this.props.isReady("guide")) { return null; }
    if(!this.props.isReady("results")) { return null; }
    if(this.state.competencies.length === 0) { return null; }

    const {translate} = this.props;
    const {competencies, displayedCompetency, showExpandedIntro} = this.state;
    const {intro, readMore} = this.introduction();

    return (
      <div className={style.tabsContainer}>
        <div className={style.tabContainer}>
          <CompetencySelect
            competencies={competencies}
            displayedCompetency={displayedCompetency}
            displayCompetency={this.displayCompetency}
            tabBadge={this.tabBadge}
          />
          <ul className={style.tabs}>
            {competencies.map((competency) => (
              <CompetencyTab
                key={competency.id}
                competency={competency}
                displayedCompetency={displayedCompetency}
                displayCompetency={this.displayCompetency}
                tabBadge={this.tabBadge}
              />
            ))}
          </ul>
        </div>
        <div className={style.tabsContent}>
          <div className={style.tabContentActive}>
            <h2>{displayedCompetency.name}</h2>
            {intro}
            <p>
              <button
                type="button"
                onClick={this.toggleExpandedIntro}
              >
                {translate("read_more")}
              </button>
            </p>
            {showExpandedIntro && (<p>{readMore}</p>)}
            <hr />
            {displayedCompetency.questionSequences.map((sequence) => (
              <div key={sequence.id}>
                <h2>{sequence.name}</h2>
                <p>{translate("guide_intro")}</p>
                <p><DangerousHTML html={translate("guide_get_started_html")} tag="em" /></p>
                {sequence.questions.map((question) => (
                  <div key={question.id}>
                    <h3 id={question.order === 1 ? "traitify-question-1" : null}>{`Question ${question.order}`}</h3>
                    <p>{question.text}</p>
                    <h4>{translate("question_purpose")}</h4>
                    <div>{this.stringToListItems(question.purpose)}</div>
                    {question.adaptability && (
                      <div>
                        <h4>{translate("question_adaptability")}</h4>
                        <div>{this.stringToListItems(question.adaptability)}</div>
                      </div>
                    )}
                  </div>
                ))}
              </div>
            ))}
          </div>
        </div>
      </div>
    );
  }
}

export {Guide as Component};
export default withTraitify(Guide);
