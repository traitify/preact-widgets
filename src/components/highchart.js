import PropTypes from "prop-types";
import {Component} from "react";
import HighCharts from "highcharts";
import HighChartsReact from "highcharts-react-official";

export default class HighChart extends Component {
  static propTypes = {
    backgroundColor: PropTypes.string,
    chart: PropTypes.shape({
      type: PropTypes.string
    }).isRequired,
    donut: PropTypes.number,
    legend: PropTypes.shape({
      layout: PropTypes.string
    }),
    series: PropTypes.arrayOf(
      PropTypes.shape({
        name: PropTypes.string,
        data: PropTypes.arrayOf(
          PropTypes.shape({
            y: PropTypes.number
          })
        )
      })
    ).isRequired,
    title: PropTypes.string,
    type: PropTypes.string,
    xAxis: PropTypes.shape({
      title: PropTypes.string
    }),
    yAxis: PropTypes.shape({
      title: PropTypes.string
    })
  }
  static defaultProps = {
    backgroundColor: null,
    donut: null,
    legend: null,
    yAxis: null,
    xAxis: null,
    title: null,
    type: null
  }
  constructor(props) {
    super(props);

    this.state = {
      config: {}
    };
  }
  componentDidMount() {
    this.afterRender();
  }
  afterRender() {
    this.config = this.setConfig();
    if(this.props.type === "radar") {
      this.configForRadar();
    } else if(this.props.type === "pie") {
      this.configForPie();
    }
    if(this.props.chart) { this.configForChart(); }
    if(this.props.legend) { this.configForLegend(); }
    if(this.props.xAxis) { this.configForX(); }
    if(this.props.yAxis) { this.configForY(); }
    this.configForTooltip();
    this.setState({config: this.config});
  }
  setConfig() {
    return {
      chart: {
        type: this.props.type,
        backgroundColor: this.props.backgroundColor
      },
      title: {text: this.props.title},
      xAxis: {title: {text: ""}, crosshair: true},
      yAxis: {title: {text: ""}},
      plotOptions: {
        column: {
          pointPadding: 0.2,
          borderWidth: 0
        }
      },
      credits: {enabled: false},
      exporting: {enabled: false},
      legend: {},
      series: this.props.series
    };
  }
  configForRadar() {
    this.config.chart.type = "area";
    this.config.chart.polar = true;
  }
  configForPie() {
    this.config.chart = {
      type: "pie",
      plotBackgroundColor: null,
      plotBorderWidth: null,
      plotShadow: false,
      height: 270,
      spacing: [0, 10, 10, 0]
    };
    this.config.plotOptions.pie = {
      allowPointSelect: true,
      cursor: "pointer",
      dataLabels: {
        enabled: true,
        distance: -30,
        format: "<b>{point.percentage:.0f}%</b>",
        style: {
          color: (HighCharts.theme && HighCharts.theme.contrastTextColor) || "white",
          textShadow: "0 0 4px #000000"
        }
      }
    };
    if(this.props.donut) { this.config.series[0].innerSize = `${100 - this.props.donut}%`; }
    this.config.series[0].colorByPoint = true;
  }
  configForX() {
    Object.assign(this.config.xAxis, this.props.xAxis);
  }
  configForY() {
    Object.assign(this.config.yAxis, this.props.yAxis);
  }
  configForChart() {
    Object.assign(this.config.chart, this.props.chart);
  }
  configForLegend() {
    Object.assign(this.config.legend, this.props.legend);
  }
  configForTooltip() {
    this.config.tooltip = {};
    if(this.config.series[0].name) {
      this.config.tooltip = {
        headerFormat: "<span style=\"font-size:10px\">{point.key}</span><table>",
        pointFormat: "<tr><td style=\"color:{series.color};padding:0\">{series.name}: </td>"
          + "<td style=\"padding:0\"><b>{point.y:.0f}</b></td></tr>",
        footerFormat: "</table>",
        shared: true,
        useHTML: true
      };
    } else {
      this.config.tooltip.formatter = function() {
        let text = "";
        if(this.point.color) { text += `<span style="color:${this.point.color}">`; }
        text += this.point.name || this.point.category;
        if(this.point.color) { text += "</span>"; }
        text += `: ${this.point.y}`;
        if(this.point.description) { text += ` <br/> ${this.point.description}`; }
        return text;
      };
    }
  }
  render() {
    return (
      <HighChartsReact
        highcharts={HighCharts}
        options={this.state.config}
      />
    );
  }
}
