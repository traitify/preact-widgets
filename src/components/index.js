export Default from "./default";
export Results from "./results";
export TypeBasedResults from "./results/type-based-results";
export DimensionBasedResults from "./results/dimension-based-results";
export SlideDeck from "./slide-deck";
export Guide from "./guide";

export * as DimensionComponents from "./results/dimension-based-results/components";
export * as TypeComponents from "./results/type-based-results/components";
