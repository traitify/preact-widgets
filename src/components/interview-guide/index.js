import PropTypes from "prop-types";
import withTraitify from "lib/with-traitify";
import {Component} from "react";
import ApolloClient from "apollo-boost";
import {ApolloProvider, Query} from "react-apollo";
import query from "graphql-queries/guide-for-assessment";

const client = new ApolloClient({
  headers: {authorization: "Basic ZGhtZ3NnMWQ2MmJxcDhpb2tqaTEzZnVobWY6eA==", mode: "no-cors"},
  uri: "https://api.stag.awse.traitify.com/v1/interview_guides/graphql"
});

class InterviewGuide extends Component {
  static defaultProps = {assessment: null}
  static propTypes = {
    assessment: PropTypes.shape({assessment_type: PropTypes.string})
  }
  buildTypes(types) {
    return types.map((type) => {
      let personalityType = type.personality_type;
      personalityType = {level: personalityType.level, id: personalityType.id};
      return {personalityType};
    });
  }
  render() {
    if(!this.props.assessment) { return null; }

    const props = this.props.assessment;
    const assessment = {
      id: props.id,
      deckId: props.deck_id,
      personalityTypes: this.buildTypes(props.personality_types)
    };

    return (
      <ApolloProvider client={client}>
        <div className="interview-guide-container">
          <Query query={query} variables={{assessment}}>
            {({loading, error, data}) => {
              if(loading) return <p>Loading...</p>;
              if(error) return <p>Error :(</p>;

              console.log(data, "query data");
              return <div>testing please wait a moment</div>;
            }}
          </Query>
        </div>
      </ApolloProvider>
    );
  }
}

export {InterviewGuide as Component};
export default withTraitify(InterviewGuide);

// <Query query={query} variables={{assessment}}>
//   {({loading, error, data}) => {
//     if(loading) return <p>Loading...</p>;
//     if(error) return <p>Error :(</p>;
//
//     console.log(data, "query data");
//     return <div>testing please wait a moment</div>;
//   }}
// </Query>
