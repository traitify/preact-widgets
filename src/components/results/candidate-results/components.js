export CandidateResults from "./index";
export PersonalityArchetype from "./personality-archetype";
export PersonalityDetails from "./personality-details";
export PersonalityDimensionColumn from "./personality-dimension-column";
export PersonalityDimensionDetails from "./personality-dimension-details";
export PersonalityDimensions from "./personality-dimensions";
