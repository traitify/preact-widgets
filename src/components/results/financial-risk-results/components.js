export FinancialRiskResults from "./index";
export PersonalityArchetype from "./personality-archetype";
export PersonalityDetails from "./personality-details";
export PersonalityDimension from "./personality-dimension";
export PersonalityDimensions from "./personality-dimensions";
export PersonalityScoreBar from "./personality-score-bar";
export PersonalityTakeaways from "./personality-takeaways";
