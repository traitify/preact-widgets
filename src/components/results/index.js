import PropTypes from "prop-types";
import {Component} from "react";
import TraitifyPropTypes from "lib/helpers/prop-types";
import withTraitify from "lib/with-traitify";
import CandidateResults from "./candidate-results";
import DimensionBasedResults from "./dimension-based-results";
import FinancialRiskResults from "./financial-risk-results";
import TypeBasedResults from "./type-based-results";

class Results extends Component {
  static defaultProps = {assessment: null}
  static propTypes = {
    assessment: PropTypes.shape({
      assessment_type: PropTypes.string,
      scoring_scale: PropTypes.string
    }),
    getOption: PropTypes.func.isRequired,
    isReady: PropTypes.func.isRequired,
    ui: TraitifyPropTypes.ui.isRequired
  }
  componentDidMount() {
    this.props.ui.trigger("Results.initialized", this);
  }
  componentDidUpdate() {
    this.props.ui.trigger("Results.updated", this);
  }
  render() {
    if(!this.props.isReady("results")) { return null; }

    if(this.props.assessment.scoring_scale === "LIKERT_CUMULATIVE_POMP") {
      return <FinancialRiskResults {...this.props} />;
    } else if(this.props.assessment.assessment_type === "TYPE_BASED") {
      return <TypeBasedResults {...this.props} />;
    } else if(this.props.getOption("view") === "candidate") {
      return <CandidateResults {...this.props} />;
    } else {
      return <DimensionBasedResults {...this.props} />;
    }
  }
}

export {Results as Component};
export default withTraitify(Results);
