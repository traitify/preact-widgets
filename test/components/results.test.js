import {Component} from "components/results";
import ComponentHandler from "support/component-handler";

jest.mock("components/results/candidate-results", () => (() => (<div className="mock">Candidate</div>)));
jest.mock("components/results/dimension-based-results", () => (() => (<div className="mock">Dimension Based</div>)));
jest.mock("components/results/financial-risk-results", () => (() => (<div className="mock">Financial Risk</div>)));
jest.mock("components/results/type-based-results", () => (() => (<div className="mock">Type Based</div>)));
jest.mock("lib/with-traitify");

describe("Results", () => {
  let props;

  beforeEach(() => {
    props = {
      isReady: jest.fn().mockName("isReady").mockReturnValue(false),
      getOption: jest.fn().mockName("getOption"),
      ui: {
        current: {},
        off: jest.fn().mockName("off"),
        on: jest.fn().mockName("on"),
        trigger: jest.fn().mockName("trigger")
      }
    };
  });

  describe("callbacks", () => {
    it("triggers initialization", () => {
      const component = new ComponentHandler(<Component {...props} />);

      expect(props.ui.trigger).toHaveBeenCalledWith("Results.initialized", component.instance);
    });

    it("triggers update", () => {
      const component = new ComponentHandler(<Component {...props} />);
      component.updateProps();

      expect(props.ui.trigger).toHaveBeenCalledWith("Results.updated", component.instance);
    });
  });

  it("renders candidate results", () => {
    props.assessment = {assessment_type: "DIMENSION_BASED"};
    props.getOption.mockReturnValue("candidate");
    props.isReady.mockReturnValue(true);
    const component = new ComponentHandler(<Component {...props} />);

    expect(component.tree).toMatchSnapshot();
  });

  it("renders dimension based results", () => {
    props.assessment = {assessment_type: "DIMENSION_BASED"};
    props.isReady.mockReturnValue(true);
    const component = new ComponentHandler(<Component {...props} />);

    expect(component.tree).toMatchSnapshot();
  });

  it("renders financial risk results", () => {
    props.assessment = {assessment_type: "DIMENSION_BASED", scoring_scale: "LIKERT_CUMULATIVE_POMP"};
    props.isReady.mockReturnValue(true);
    const component = new ComponentHandler(<Component {...props} />);

    expect(component.tree).toMatchSnapshot();
  });

  it("renders type based results", () => {
    props.assessment = {assessment_type: "TYPE_BASED"};
    props.isReady.mockReturnValue(true);
    const component = new ComponentHandler(<Component {...props} />);

    expect(component.tree).toMatchSnapshot();
  });

  it("renders nothing if not ready", () => {
    const component = new ComponentHandler(<Component {...props} />);

    expect(component.tree).toMatchSnapshot();
  });
});
