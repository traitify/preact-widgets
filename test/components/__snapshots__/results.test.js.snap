// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Results renders candidate results 1`] = `
<div
  className="mock"
>
  Candidate
</div>
`;

exports[`Results renders dimension based results 1`] = `
<div
  className="mock"
>
  Dimension Based
</div>
`;

exports[`Results renders financial risk results 1`] = `
<div
  className="mock"
>
  Financial Risk
</div>
`;

exports[`Results renders nothing if not ready 1`] = `null`;

exports[`Results renders type based results 1`] = `
<div
  className="mock"
>
  Type Based
</div>
`;
